src/circuit/circuit_frame.rs

```diff
@@ -2278,30 +2288,78 @@ fn apply_continuation<F: LurkField, CS: ConstraintSystem<F>>(
             store,
         )?;
 
+        let op_is_begin = alloc_equal(
+            cs.namespace(|| "op_is_begin"),
+            operator.tag(),
+            &g.op2_begin_tag,
+        )?;
+
         let rest_is_nil =
-            allocated_rest.alloc_equal(&mut cs.namespace(|| "args_is_nil"), &g.nil_ptr)?;
+            allocated_rest.alloc_equal(&mut cs.namespace(|| "rest_is_nil"), &g.nil_ptr)?;
 
-        let the_cont = AllocatedContPtr::pick(
-            &mut cs.namespace(|| "the_cont"),
+        let begin = store.get_begin();
+
+        let allocated_begin =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "begin"), store, || Ok(&begin))?;
+
+        let begin_again = AllocatedPtr::construct_cons(
+            &mut cs.namespace(|| "begin again"),
+            g,
+            store,
+            &allocated_begin,
+            &unevaled_args,
+        )?;
+
+        let the_expr_if_begin = AllocatedPtr::pick(
+            &mut cs.namespace(|| "the_exp_if_begin"),
             &rest_is_nil,
-            &newer_cont,
-            &g.error_ptr_cont,
+            &allocated_arg2,
+            &begin_again,
         )?;
 
-        let the_expr = AllocatedPtr::pick(
-            &mut cs.namespace(|| "the_expr"),
+        let otherwise = Boolean::not(&op_is_begin);
+
+        let otherwise_if_rest_is_nil = Boolean::and(
+            &mut cs.namespace(|| "otherwise_if_rest_is_nil"),
+            &otherwise,
+            &rest_is_nil,
+        )?;
+
+        let the_expr_otherwise = AllocatedPtr::pick(
+            &mut cs.namespace(|| "the_expr_otherwise"),
             &rest_is_nil,
             &allocated_arg2,
             result,
         )?;
 
+        let the_expr = AllocatedPtr::pick(
+            &mut cs.namespace(|| "the_expr"),
+            &otherwise,
+            &the_expr_otherwise,
+            &the_expr_if_begin,
```

Comment:
I think you can eliminate one of the three `pick`s you are using to select `the_expr`. The observation is that `arg2` appears in both top-level branches, as structured.

Instead, you could have something structurally like this pseudocode:
```
let the_expr = 
  if rest_is_nil { arg2 }
  else { if op_is_begin { begin_again }
            else { result } }
```